python/events/src/hydrotools/events/baseflow/eckhardt.py

```diff
@@ -0,0 +1,288 @@
+        maximum_baseflow_index: float
+            The maximum baseflow index, $BFI_{max}$, from Eckhardt (2005, 2008). Must be
+            between 0.0 and 1.0, typically between 0.25 and 0.8 for daily streamflow.
+            Influenced by aquifer porosity, streamflow frequency, and other catchment
+            characteristics.
+            
+        Returns
+        -------
+        baseflow: array-type
+            An array containing the separated baseflow values.
+    """
+    # Compute filter parameters
+    denominator = 1 - recession_constant * maximum_baseflow_index
+    A = (1.0 - maximum_baseflow_index) * recession_constant / denominator
+    B = (1.0 - recession_constant) * maximum_baseflow_index / denominator
+
+    # Instantiate baseflow series
+    # Assume first value is baseflow
+    baseflow = np.empty(series.size)
+    baseflow[0] = series[0]
+
+    # Apply filter and return result
+    for i in range(1, series.size):
+        baseflow[i] = min(series[i], A * baseflow[i-1] + B * series[i])
+    return baseflow
+
+def separate_baseflow(
+        series: pd.Series,
+        output_time_scale: Union[pd.Timedelta, datetime.timedelta, np.timedelta64, str, int],
+        recession_time_scale: Union[pd.Timedelta, datetime.timedelta, np.timedelta64, str, int] = "1D",
+        recession_window: int = 5
+) -> BaseflowData:
+    """Applies a digitial recursive baseflow separation filter to series
+        and returns the result. This is a higher level method than `apply_filter`
+        that includes time resampling and filter parameter estimation.
+    
+        Parameters
+        ----------
+        series: pandas.Series, required
+            A pandas.Series of streamflow values with a DateTimeIndex. Assumes
+            first and last values in series are baseflow.
+        output_time_scale: pandas.Timedelta, datetime.timedelta, numpy.timedelta64, str, int, required
+            Output time-scale (or 'time-step') of the output baseflow time series.
+            Typically the same time-scale as series.
+        recession_time_scale: pandas.Timedelta, datetime.timedelta, numpy.timedelta64, str, int, optional, default '1D'
+            Time-scale or 'time-step' over which to conduct recession analysis.
+            Generally, it's better to use a recession_time_scale greater than the series time-scale.
+            The default of '1D' is good for many catchments, but smaller catchments can benefit from a
+            shorter time-scale of '18h' or '12h'.
+        recession_window: int, optional, default 5
+            The minimum number of consecutively decreasing values in series
+            that indicate a period of recession.
+
+        Returns
+        -------
+        baseflow_data: BaseflowData
+            A BaseflowData DataClass containing the separated baseflow and associated filter parameters.
+    """
+    # Compute recession constant
+    recession_series = series.resample(recession_time_scale).nearest(limit=1)
```

Comment:
This should be mean, not nearest (thanks Austin!)

Reply:
This actually needs to be updated to allow a flexible aggregation with some guidance. To duplicate Eckhardt's method using daily mean, it should be mean. However, when using this method with discharge per unit area (AKA "runoff") it should use sum. There might also be cases where you'd want to use to "nearest" or "first" with instantaneous data. I'll refactor this to use `agg` and add a parameter.